crates/staged-sync/src/utils/init.rs

```diff
@@ -22,11 +22,20 @@ pub fn init_db<P: AsRef<Path>>(path: P) -> eyre::Result<Env<WriteMap>> {
 }
 
 /// Write the genesis block if it has not already been written
+///
+/// # Panics
+///
+/// Panics if the database has already been initialized with a different genesis block.
 #[allow(clippy::field_reassign_with_default)]
 pub fn init_genesis<DB: Database>(db: Arc<DB>, chain: ChainSpec) -> Result<H256, reth_db::Error> {
     let genesis = chain.genesis();
+
+    let header = chain.genesis_header();
+    let hash = header.hash_slow();
+
     let tx = db.tx()?;
-    if let Some((_, hash)) = tx.cursor_read::<tables::CanonicalHeaders>()?.first()? {
+    if let Some((_, db_hash)) = tx.cursor_read::<tables::CanonicalHeaders>()?.first()? {
+        assert_eq!(hash, db_hash, "Genesis hash mismatch: expected {}, got {}", hash, db_hash);
```

Comment:
I'd be ok with adding a custom error for this, perhaps even a new error type just for this.

perhaps something simple like `enum InitDatabaseError {...}`